scripts/lint_jjb.py

```diff
@@ -316,16 +317,18 @@ def check_timed_trigger(data, in_file):
 if __name__ == "__main__":
     args = parse_args()
     rc = 0
+
+    excludes = map(operator.methodcaller("strip"),
+                   args.exclude_files.split(','))
     for _dir in args.dirs.split(','):
-        for root, dirs, files in os.walk(_dir):
-            skip_dir = False
-            excludes = map(lambda s: s.strip(), args.exclude_files.split(','))
-            for exclude in excludes:
-                if "{s}{ex}".format(s=os.sep, ex=exclude) in root:
-                    skip_dir = True
-                    break
-            if skip_dir:
-                continue
+        walk_list = list(os.walk(_dir))
```

Comment:
By constructing the list of all files and directories here, it removes the benefit of modifying the directory list and no longer skips files in excluded directories. It's now also only trying to skip them at the top level, is that intentional? I've added a little example below which I think shows the difference.

```
$ ls -1R foo
a
b
c

foo/a:
a.txt

foo/b:
b.txt

foo/c:
c.txt
```

```python
import os                                                                          
                                                                                   
excludes = ["b"]                                                                   
_dir = "/tmp/foo"                                                                  
                                                                                   
walk_list = list(os.walk(_dir))                                                    
_, dirs, _ = walk_list[0]                                                          
for exclude in excludes:                                                           
    try:                                                                           
        dirs.remove(exclude)                                                       
    except ValueError:                                                             
        pass                                                                       
                                                                                   
print("What this code is doing:")                                                  
print("\n".join(str(item) for item in walk_list))                                  
                                                                                   
                                                                                   
new_walk_list = []                                                                 
for root, dirs, files in os.walk(_dir):                                            
    for exclude in excludes:                                                       
        try:                                                                       
            dirs.remove(exclude)                                                   
        except ValueError:                                                         
            pass                                                                   
    new_walk_list.append((root, dirs, files))                                      
                                                                                   
print("\nWhat this code should be doing:")                                         
print("\n".join(str(item) for item in new_walk_list))
```

```
What this code is doing:
('/tmp/foo', ['a', 'c'], [])
('/tmp/foo/a', [], ['a.txt'])
('/tmp/foo/c', [], ['c.txt'])
('/tmp/foo/b', [], ['b.txt'])

What this code should be doing:
('/tmp/foo', ['a', 'c'], [])
('/tmp/foo/a', [], ['a.txt'])
('/tmp/foo/c', [], ['c.txt'])
```

Reply:
@git-harry thanks for the clear explanation, I see that the dirs objects in each tuple are independent so my method of only removing it once won't work. I was assuming some reference object, but thats not the case with a string.